### ISSUE TITLE
storage: proposal blocked on replica no longer part of range

### ISSUE THREAD

**Issue by petermattis** [cockroachdb/cockroach#15551]
While debugging #15163, I noticed a `replica send` operation that has been in progress for 890s:

```
16:25:31.840691	890.349519	replica send
16:25:31.840700	 .     9	... [n1,s1,r3049/18:/Table/51/1/60{88103…-90365…}] read-write path
16:25:31.840732	 .    32	... [n1,s1,r3049/18:/Table/51/1/60{88103…-90365…}] raft
16:26:31.903690	60.062958	... storage/replica.go:2262 [n1,s1,r3049/18:/Table/51/1/60{88103…-90365…}] have been waiting 1m0s for proposing command RequestLease [/Table/51/1/6088103647699277859/"54053aeb-52b1-494f-a364-44e2b5f9cef1"/1196428,/Min)
```

Looking at the `debug/range?id=3049` I can see that the range currently has replica IDs of `19`, `20` and `21`. I'm going to dive into the logs to see if I can figure out when replica ID `18` (the replica the proposal is stuck on) was removed. And why this node didn't cancel the pending proposal. 

Cc @bdarnell

**Comment by petermattis** [comment]
```
I170501 16:24:55.127695 527 storage/store.go:3149  [n1,s1,r3049/18:/Table/51/1/60{88103…-90365…}] added to replica GC queue (peer suggestion)
W170501 16:25:31.840664 665629 storage/node_liveness.go:291  [n1,s1,r3049/18:/Table/51/1/60{88103…-90365…}] slow heartbeat took 2.9s
I170501 16:25:31.850195 91 storage/store.go:2137  [replicaGC,n1,s1,r3049/18:/Table/51/1/60{88103…-90365…}] removing replica
I170501 16:25:31.901654 91 storage/replica.go:673  [replicaGC,n1,s1,r3049/18:/Table/51/1/60{88103…-90365…}] removed 466888 (466878+10) keys in 51ms [clear=14ms commit=37ms]
```

The Raft proposal was nearly concurrent with the Replica being removed. Note that the "raft" event occurs before the proposal. 

@bdarnell It looks like we don't check `Replica.mu.destroyed` atomically with submitting the proposal in `Replica.propose`. Specifically, we check `Replica.IsDestroyed` at the start of the method, but I think we need to check it again after grabbing `Replica.mu`. Tossing in an additional check would be easy and low-risk. Should we do this for 1.0?

**Comment by bdarnell** [comment]
Yeah, if that's all the fix needs then we can do it for 1.0.
